### DISCUSSION TITLE
"Cross-Origin Request Blocked" by sending request from TypeScript REST Client(s) in Angular

### DISCUSSION THREAD

**Discussion by dpedwards** [Budibase/budibase#6051] [Help] [answered]
Hi dear BB dev team, we are trying to address the Budibase CRUD API in Angular. Unfortunately without success so far, although we use three different REST Client implementations.

TypeScript REST Client #1:
![RESTClient1](https://user-images.githubusercontent.com/2211088/170130150-078721a5-1ebf-444a-b33b-ed4d03c9a3a3.png)

TypeScript REST Client #2:
![RESTClient2](https://user-images.githubusercontent.com/2211088/170130154-19e4f41a-547d-4575-bf3d-c2bd43b6241a.png)

TypeScript REST Client #3:
![RESTClient3](https://user-images.githubusercontent.com/2211088/170130166-7f8f42c0-a3e2-43d9-9e50-a63e1b428cc2.png)

**The error messages are the same in all three cases:**
![Corss-Origin_ErrorFullDetails](https://user-images.githubusercontent.com/2211088/170133931-1255f67e-58dd-4f6a-80df-56e34d92ce09.png)

![Corss-Origin_ErrorDetails](https://user-images.githubusercontent.com/2211088/170130069-ceccb468-e8bf-46b1-9cd2-d3f555ccd7d1.png)

However, with the Postman tool and the same header information as the three previously listed REST TypeScript client implementations, a successful request is possible.
Postman:
![Postman](https://user-images.githubusercontent.com/2211088/170130097-0fccc7d0-433f-47fb-8dac-0496974d1fe2.png)

So what are we doing wrong?

**Comment by melohagan** [comment, marked as answer]
Hi @dpedwards 

If I understand this correctly, we would need to send back the mentioned header: `Access-Control-Allow-Origin` in the API response.

However, from this StackOverflow post, it does look like you can use a proxy through the angular-cli to get around this: https://stackoverflow.com/questions/50237881/angular-cors-request-blocked

Out of curiosity, does this issue persist across different web browsers?

**Reply by dpedwards** [reply]
"_If I understand this correctly, we would need to send back the mentioned header: Access-Control-Allow-Origin in the API response._"

Hi, 
that would probably help.

![2022-05-24 23_10_56-Angular – Access to XMLHttpRequest at “http___… ” origin ‘http___localhost_4200’](https://user-images.githubusercontent.com/2211088/170214920-c51401d2-59c9-438b-a4f2-f5e488d254f5.png)

[Source](https://itecnote.com/tecnote/angular-access-to-xmlhttprequest-at-http-origin-http-localhost4200-has-been-blocked-by-cors-policy/)

**Reply by dpedwards** [reply]
"_However, from this StackOverflow post, it does look like you can use a proxy through the angular-cli to get around this: https://stackoverflow.com/questions/50237881/angular-cors-request-blocked_"

Thanks for the link, we'll check it out.

"_Out of curiosity, does this issue persist across different web browsers?_"

Whether using custom REST clients, or viewing the Angular WebApp in Firefox, Chrome, or Edge, the cross-origin request error message is always shown. Only with Postman we can send requests to the BB CRUD API.

**Reply by dpedwards** [reply]
This thread can be closed.

Btw if you use an API (e.g. Python FastAPI) as wrapper and send the request from an Angular service over it to the BB CRUD API then the request from an Angular component also works locally. The [proxy](https://stackoverflow.com/questions/50237881/angular-cors-request-blocked), however, did not work in Angular.

**Comment by Dharaneesh0745** [comment]
Hii @dpedwards 

If you're using Typescript and facing a CORS error in production means, just write the ORIGIN as usual in your code. Before pushing your code to production, don't forget to run the build command, `npm run build`. Because in production, it always runs the Javascript code not the Typescript code. Once you have built your code using the build command, where the build Javascript files have been stored inside a dist or build folder based on your `tsconfig.json` configuration. On every time if you made changes in typescript code, don't forget to run the build command before pushing your code to production environment.

Sample code of mine,

```
const allowedOrigins = ["http://localhost:3000", "https://mudomain.com"]

app.use(
  cors({
    origin: (origin, callback) => {
      if (!origin || allowedOrigins.includes(origin)) {
        callback(null, true);
      } else {
        callback(new Error("Your client url is not allowed by CORS"))
      }
    },
    credentials: true
  })
)
```

![image](https://github.com/user-attachments/assets/27fe4df8-2155-4292-95e4-470edd17fce8)

**Reply by dpedwards** [reply]
@Dharaneesh0745, thanks for the final solution! By the way, we solved the CORS issue by creating an API wrapper using FastAPI. We handled the Angular requests through the API wrapper and routed them to the Budibase API.
